### ISSUE TITLE
MEC: Be lens calculations.

### ISSUE THREAD

**Issue by silkenelson** [pcdshub/pcdscalc#1]
**Feature Request**
- calcBeamFWHM
- calcFocalLength
- calcFocalLengthForSingleLens
- calcLensApertureRadius
- calcTransForSingleLens
- calcTransLensSet

<Zach comment>
Most of the calcs landed into the lens classes in pcdsdevices but it could use a once-over

Maybe we just need documentation in confluence about how to use the BeLens class?

**Comment by mseaberg** [comment]
I think the old code was mostly fine. The main thing to add was the finite source distance for planning lens sets. Previously the calculation assumed infinite source distance which in practice meant people usually had to move the beamline lenses upstream to get the focus where they wanted it. This can be fixed by adding one line to the calcBeamFWHM function, along with an additional keyword argument to that function. I am happy to tell you what to add, or I can contribute directly to the code if that's easier. There's may also still be a couple more functions to transfer over to recover what was available in xpppython.

**Comment by cristinasewell** [comment]
Thanks @mseaberg for looking into this - I have added a PR with the calculations (please note it is a work in progress, needs lots of cleanup). I have the `calc_beam_fwhm` function here - please feel free to comment/suggest the additional line and keyword argument that you mentioned above in the PR:  https://github.com/pcdshub/pcdscalc/pull/5 - and thanks a lot
Here is the function: 
https://github.com/pcdshub/pcdscalc/blob/3175317fc52994b386791292214cf63084b26205/pcdscalc/be_lens_calcs.py#L118-L167

Also, you mentioned a couple more functions to transfer over, are you talking about these (link below), if so, do I need to change anything about them? 
https://github.com/pcdshub/old-hutch-configs/blob/9110538b0f8119ed6f8960a1ffd24191ed798178/mec/common/utilitiesBeLenses.py#L158-L227
If you were not talking about these, could you please point me to the ones that we want to transfer over?

**Comment by mseaberg** [comment]
I made a suggestion for the source_distance argument and additional lines. For the additional functions the main one I had in mind was planSet from here that wraps calcLensSet (I have not used calcLensSet directly before, can look into the differences): [https://github.com/pcdshub/old-hutch-configs/blob/master/xpp/blinst/crl.py#L414](https://github.com/pcdshub/old-hutch-configs/blob/master/xpp/blinst/crl.py#L414). There are others in this file that I guess generate a lookup table for the x and y centers as the lenses are moved along the z stage. I haven't used this before but I think Diling was asking about adding something like this into xpp3. This is almost like a virtual motor, not sure if this would be better to write from scratch or try to re-use. Also I think it would belong somewhere else since this is not a calculation tool but something for actually moving motors.

**Comment by cristinasewell** [comment]
That's great, hanks @mseaberg  - I will implement the `planSet` then :) 
As of the suggestion for `source_distance` - I don't know how to find the suggestion, can you please direct me where to look for it?

**Comment by mseaberg** [comment]
I put it straight into your PR (I think). I am still a bit new to github, I may have done it wrong or may have misunderstood what you were asking for. If you can't see it there I'll try again later.

**Comment by cristinasewell** [comment]
> I put it straight into your PR (I think). I am still a bit new to github, I may have done it wrong or may have misunderstood what you were asking for. If you can't see it there I'll try again later.

Yes that's exactly what I was thinking when I asked about your suggestion but I don't see it in the PR (at least not yet). I have looked here for it:  https://github.com/pcdshub/pcdscalc/pull/5/commits/3175317fc52994b386791292214cf63084b26205

If it is more convenient for you to just write it here in the comments that would work great for me too - thank you very much for your help with this.

**Comment by klauer** [comment]
You may have started a draft review, @mseaberg, and it may just be that you didn't submit it yet. If you go back to the PR, it should show an option to submit the comments somewhere near the bottom.

**Comment by mseaberg** [comment]
Yes, @klauer, that was the problem. Thanks! @cristinasewell, I think it should show up now.

**Comment by cristinasewell** [comment]
Thanks @mseaberg  - yes I see it now :)

**Comment by silkenelson** [comment]
> I made a suggestion for the source_distance argument and additional lines. For the additional functions the main one I had in mind was planSet from here that wraps calcLensSet (I have not used calcLensSet directly before, can look into the differences): https://github.com/pcdshub/old-hutch-configs/blob/master/xpp/blinst/crl.py#L414. There are others in this file that I guess generate a lookup table for the x and y centers as the lenses are moved along the z stage. I haven't used this before but I think Diling was asking about adding something like this into xpp3. This is almost like a virtual motor, not sure if this would be better to write from scratch or try to re-use. Also I think it would belong somewhere else since this is not a calculation tool but something for actually moving motors.

This virtual motor will/does exist in the XFLS class. This issue is mostly to move the calculation classes into their own package & documet how to use them and use these classes in the devices themselves (pre-focus, XFLS/CRL, transfocator [maybe o that one]).

**Comment by cristinasewell** [comment]
So the `lens` Classes in the new code don't seem to have these function:
```
calcLensApertureRadius
calcTransForSingleLens
calcTransLensSet
```
unless they are in a different place that I don't know or renamed, but regardless, do we still need these functions? (i'm assuming we do if they were part of **Feature Request** but not sure )

**Comment by silkenelson** [comment]
The function request came from MEC who has been using the same code, but might have developed it more and/or used an older version. I'll ping Eric Galtier to look at this request.

**Comment by cristinasewell** [comment]
> The function request came from MEC who has been using the same code, but might have developed it more and/or used an older version. I'll ping Eric Galtier to look at this request.

Thank you very much ☺️

**Comment by egaltier** [comment]
We would definitely need to get the following :
```
calcLensApertureRadius
calcTransForSingleLens
calcTransLensSet
```
as these would allow us to estimate the total transmission of the lenses (it is of importance to optimize which lens radius to use at a specific photon energy).

From the thread it is not clear if you have already the following methods:
```
setLensSetsToFile
getLensSet
```
The Be lens holders can take 3 different sets that we usually set before few experiments so we only vent the relevant beamline section once. We then store these sets into a config file that we can call from the methods you redefined. Later, we save this config file to the specific experiment so users can make sure that they know which stack was used for there beamtime. Is that something we can get as well?

Finally, a final request would be to allow for extended photon energy coverage. For now, the values of the attenuation length from the `getAttLen` method and of delta (1-index of refraction) from `getDelta` are taken from the python module `periodictable`. The tables used stop at 30 keV. Future projects will need to enable up to 45 keV photon energies (e.g. the third harmonics of 15 keV), so the `periodictable` would not work anymore. I have been using the `xraydb` package (imported as `xdb`) to extend the value to arbitrary photon energies. I have made a comparison plot between the `periodictable`, the `xraydb` package, the CXRO database, and the calculation produced by the `shadow` code for Be and Gold, on both the attenuation length and delta.
![att_len_2-45keV](https://user-images.githubusercontent.com/44449282/95420394-320fc680-08f0-11eb-8cdd-9b60d98daf7b.png)
![delta_2-45keV](https://user-images.githubusercontent.com/44449282/95420392-2fad6c80-08f0-11eb-8e5a-883ac9282bef.png)
For the attenuation length, `xraydb` and `shadow` show very similar results with each other in both cases. They agree well to the CXRO/NIST database on gold while the `periodictable` for Be seems very off at high energies. For delta, all approaches show very similar results. Essentially, this would show that it would be good to switch table and use the `xraydb` module instead of the `periodictable`, especially since we can go to arbitrarily high photon energies while keeping (and improving) the accuracy of the calculation for the CRL. In terms of code, it would look like that:
```
import xraydb as xdb

def getAttLen(self, E, material, density=None):
     ```
     E is given in keV
     material is a string e.g. 'Be'
     ```
     att_len = 1. / xdb.material_mu(material, E * 1.0e3)
     return att_len

def getDelta(self, E, material="Be", density=None):
     ```
     E is given in keV
     material is a string e.g. 'Be'
     ```
     delta = xdb.xray_delta_beta(material, energy=E * 1.0e3, density=xdb.atomic_density(material))[0]
     return delta
```
I hope this makes sense.

**Comment by silkenelson** [comment]
Hi Eric, do you need these functions explicitly or do you need the functionality to 'plan' a set of lenses to use for a given focus at a chosen beam energy? That the current set of code has. Good points on making sure our code can cover the future higher energies and to point out that we need a way to store & retrieve sets.

**Comment by egaltier** [comment]
If I understand your question correctly (not sure here :p) I believe both. To plan for the right lens set for an experiment, we basically do some trial and error with the calcBeamFWHM to estimate how many lenses could be used to reach a range of spot at a specific photon energy. Also, sometime during a beamtime, we check the transmission of the stacks to understand some of the signal we see, so within the python interactive environment, we get the individual methods to get the lens aperture and other quantities.

**Comment by cristinasewell** [comment]
That's awesome! Thanks a lot for the information/explanation @egaltier - I'll add all those functions in, and I'll switch to trying to use `xraydb` 😊 - thanks for the code above
As for these: 
```
setLensSetsToFile
getLensSet
```
I think we have them in the new `LensStack` class but I will add them here too, since we use them in the calculations. 

Thank you

**Comment by cristinasewell** [comment]
@egaltier I am trying out the `xraydb` module for the `get_delta` function, and I see that there is a bit of a difference between the results with the implementation with `xraydb` and `periodictable` and I would like to check that with with you. (it might be that with `xraydb` we get more accurate results?? )

Here is a snippet of my testing code:
```
import xraydb as xdb

from periodictable import xsf

def get_att_len(energy, material="Be", density=None):

    delta = xdb.xray_delta_beta(material, energy=energy * 1.0e3,
                                density=xdb.atomic_density(material))[0]

    old_delta = 1 - np.real(xsf.index_of_refraction(material,
                            density=xdb.atomic_density(material),
                            energy=energy))
    print(xdb.atomic_density(material))
    return delta, old_delta
```
I am using the `xdb.atomic_density(material)` for the old_delta calculation for now (just for testing), to make sure I am using the same density (i think it is just 1.85) . Also, only using `Be` to test for now for the material. 

Results:
```
At energy: 8, Received (5.330471261281744e-06, 5.332219193743271e-06)
At energy: 9, Received (4.21084077002856e-06, 4.212123752345143e-06)
At energy: 1, Received (0.0003515107795702183, 0.00035281288547495127)
At energy: 0.1, Received (0.005621922785393444, 0.0038114671984230686)
At energy: 20, Received (8.521078856544226e-07, 8.523017026584867e-07)
```

Is there a way I could know what values are more accurate? Does this difference matter?

(I could also be missing something here or maybe not testing correctly, I am not sure)

**Comment by mseaberg** [comment]
@cristinasewell these are close enough to the same.

**Comment by egaltier** [comment]
@cristinasewell Yes these are ok. I plotted the delta values across 2 to 45 keV range in my previous post and the values were very close to each other, especially on Be. The reference we use is from the CXRO. We don't really use Be lenses for photon energies below 2-3 keV ish, otherwise the absorption in the material is too high. KB mirrors are the preferred solution in this case.

**Comment by cristinasewell** [comment]
@egaltier I'm also getting different result for the `get_att_len` function, and I would like to check that with you as well when you get a chance, this time they differ a bit much from the old ones.

 
I have posted some results where I compare the old values with the new ones in the link below. The new ones are using the new `get_att_len` and `get_delta`.

https://github.com/pcdshub/pcdscalc/pull/5#issuecomment-706449344

Please let me know if you can't access the comment link above. 

I am using the following functions for the old code tests (maybe I am using the wrong one?): 
```
def getAttLen(energy, material="Be", density=None):
    att_len = float(attenuation_length(material, density=density, energy=energy))
    return att_len



def attenuation_length(
    compound, density=None, natural_density=None, energy=None, wavelength=None
):
    """
    Calculates the attenuation length for a compound
		Transmisison if then exp(-thickness/attenuation_length)
    :Parameters:
        *compound* : Formula initializer
            Chemical formula.
        *density* : float | |g/cm^3|
            Mass density of the compound, or None for default.
        *natural_density* : float | |g/cm^3|
            Mass density of the compound at naturally occurring isotope abundance.
        *wavelength* : float or vector | |Ang|
            Wavelength of the X-ray.
        *energy* : float or vector | keV
            Energy of the X-ray, if *wavelength* is not specified.
    :Returns:
        *attenuation_length* : vector | |m|
            as function of (energy)
    :Notes:
    against http://henke.lbl.gov/optical_constants/
    """
    if energy is not None:
        wavelength = xsf.xray_wavelength(energy)
    assert wavelength is not None, "scattering calculation needs energy or wavelength"
    if np.isscalar(wavelength):
        wavelength = np.array([wavelength])
    n = xsf.index_of_refraction(
        compound=compound,
        density=density,
        natural_density=natural_density,
        wavelength=wavelength,
    )
    attenuation_length = (wavelength * 1e-10) / (4 * np.pi * np.imag(n))
    return np.abs(attenuation_length)

```

**Comment by egaltier** [comment]
@cristinasewell I got the same results than you, for both the old and new values so your methods are right. It is just that the new values are given in cm while the old in m. So, when you compare the new values to the old one, you should multiply the new values by `1.0e-2`. After this is done, the discrepancies between the old and new values are correct. From the graph I made 2 days ago on the attenuation length calculation over the entire photon energy range, the new values are closer to the CXRO reference, so this is an improvement vs our previous method.

**Comment by cristinasewell** [comment]
@egaltier that's great, thanks so much for letting me know about that, and thanks a lot for taking the time too look at the results and test them, I appreciate it ☺️

**Comment by egaltier** [comment]
No problemo @cristinasewell ! Thanks a lot to invest your time in improving the LCLS python capabilities! It makes a big difference during preparation and operation :)

**Comment by cristinasewell** [comment]
I should have asked this a while ago but are we going to need these functions below as well? They were in the old code part of `BeLens` calculations and I added them here, but I would like to take them off if we don't use these anymore:

`find_radius`
`find_energy`
`find_z_pos`

https://github.com/pcdshub/pcdscalc/blob/d0b782ce5644eaa3828d85270912ab33617140ce/pcdscalc/be_lens_calcs.py#L699-L830

**Comment by egaltier** [comment]
These functions provide the positions of the lens stack that would bring best focus in between two upstream and downstream values. These z positions are used into a subsequent function that will move the Be lens stack motor to the desired middle position. It was meant to do some automation in the Be lens position to get to a desired spot size. By setting the new photon energy, the command was moving the lens stack to get back to best focus. It would be very much appreciate if we could still have them since it would prevent errors during operation.

**Comment by cristinasewell** [comment]
Sounds great, thank you!

**Comment by cristinasewell** [comment]
There is one more thing I would like to revise with anybody that knows. I would like to see if I should keep these defaults or if they need to be changed, and where could I get the default information from if I should change them. We are mainly using these defaults: 
```
LENS_RADII = [50e-6, 100e-6, 200e-6, 300e-6, 500e-6, 1000e-6, 1500e-6]
material='Be'
fwhm_unfocused=500e-6		  
disk_thickness=1.0e-3
apex_distance=30e-6			   
n_max=12
max_each=5
lens_radii=[100e-6, 200e-6, 300e-6, 500e-6, 1000e-6]
fwhm_unfocused=0.0005 (in a different function)
```
we're using them in multiple functions, but just pasting here a couple for reference. I am assuming we'll be using the same defaults in all the functions? or could they be different depending on the function?
```
def calc_trans_for_single_lens(energy, radius, material='Be', density=None,
                               fwhm_unfocused=500e-6, disk_thickness=1.0e-3,
                               apex_distance=30e-6)
calc_lens_set(energy, size_fwhm, distance, n_max=12, max_each=5,
                  lens_radii=[100e-6, 200e-6, 300e-6, 500e-6, 1000e-6],
				   fwhm_unfocused=0.0005, eff_rad0=None):
```

**Comment by egaltier** [comment]
The various lens stack radius of curvature we used would look like
```
LENS_RADII = [50e-6, 100e-6, 200e-6, 300e-6, 500e-6, 1000e-6, 1500e-6,  2000e-6,  3000e-6]
```
I woud like to be able to give you a link of the inventory of the lenses we have at LCLS but I don't know it, maybe @mseaberg has access to it.
For the unfocused FWHM, it is a bit more tricky since the beam has some divergence in it. Essentially, where it is 400 mic at XPP it is usually 800 mic few 100 m downstream at MEC. So on the MEC side we used 800 mic as FWHM. Also, this beam side is a function of the photon energy (the higher the photon energy the small the divergence, the smaller the beam size far downstream). I guess, this being said, since this is a default value, I would just keep 500 mic ^^; Sorry for the extra details and not going to the point. The rest of the numbers are all fine as is. I would bump the `n_max` to 25 lenses though since sometimes, for tight focus, we do used those directly in the target chamber so we are not constrained by the hutch lens holder.
